File: JavaScript/프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.js
function solution(s) {
    let answer = 0;
    for(let x = 0; x < s.length; x++){
        let newS = s.slice(x) + s.slice(0,x);
        if(isValid(newS)){
            answer ++;
        }
    }
    return answer;
}

function isValid(str) {
    let stack = [];
    let match = {")" : "(", "]" : "[", "}" : "{"};
    for(const char of str){
        if(char === "(" || char === "[" || char === "{"){
            stack.push(char);
        }
        if(char === ")" || char === "]" || char === "}"){
            if(stack.length === 0 || stack.pop() !== match[char]){
                return false;
            }
        }
    }
    return stack.length === 0;
    
}

File: Python3/프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
from collections import deque

def is_valid(s):
    stack = deque()
    for ch in s:
        if ch in '([{':
            stack.append(ch)  
        else:
            if not stack:
                return False
            top = stack.pop()
            if (top == '(' and ch != ')') or \
               (top == '{' and ch != '}') or \
               (top == '[' and ch != ']'):
                return False
    return not stack
    
def solution(s):
    n = len(s)
    answer = 0
    double_s = s + s
    
    for i in range(n):
        rotated = double_s[i:i+n]
        if is_valid(rotated):
            answer += 1
    
    
    
    return answer